keps/sig-network/4559-kubelet-probes/README.md

```diff
@@ -0,0 +1,1037 @@
+   exec probes, then this implies that CRI will need to understand
+   each type of probe supported by Kubernetes (i.e., it will need to
+   support every feature of `HTTPGetAction`, `TCPSocketAction` and
+   `GRPCAction`.) In the future, adding new probe types (for example,
+   a UDP probe) would require adding support to both `Pod`/kubelet and
+   to CRI.
+
+     - If CRI is going to need to understand that much of the
+       semantics of Kubernetes probes, then it might make sense to
+       just _completely_ move probes into CRI, and have CRI simply
+       report back as part of the pod status whether the pod is live,
+       ready, etc, according to its probes. This might allow for
+       better optimization of probes on the CRI side.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+#### Goroutines vs. Network Namespaces
+
+Moving pod probes into the pod network namespace will run into the
+standard difficulties with using kernel namespaces from golang code.
+
+The `setns()` syscall in Linux changes the namespace of a thread; this
+does not always interact well with goroutines in golang. In
+particular, you can ensure that the current goroutine won't be moved
+to another thread, and you can ([since go 1.10]) ensure that other
+goroutines won't be moved onto your thread, but you cannot (and in the
+general case would not want to) guarantee that new goroutines spawned
+from your goroutine will run in the current thread. This means that it
+is unsafe for a golang function which has called `setns()` to invoke
+any code which it is not _certain_ will perform all of its work in the
+current goroutine. In particular, [`net.Dial` does not guarantee that
+it will establish the connection from the same goroutine it was called
+from] so it is unlikely that a golang-based CRI implementation would
+be able to run probes directly from its main process.
+
+[since go 1.10]: https://github.com/golang/go/commit/2595fe7fb6
+[`net.Dial` does not guarantee that it will establish the connection from the same goroutine it was called from]: https://github.com/golang/go/issues/44922
+
+### Risks and Mitigations
+
+If we change the semantics of the existing probe types (i.e., just say
+that henceforth if you don't specify `TCPSocketAction.Host` or
+`HTTPGetAction.Host` then the probe is done in the pod network
+namespace rather than going across the network) then we risk breaking
+people who are doing sufficiently weird things. (I'm not even sure
+what these "sufficiently weird things" would be. One example would be
+having the pod recognize the node IP as a source IP so that it can
+respond differently to kubelet probes than it would to a "real"
+request.) It would be safer (though more annoying) to instead
```

Comment:
So should we say we definitely want to go with

- Add new probe types (with new semantics and no `.Host` support)
- Deprecate the old probe types
- Allow using PSA to block Pods that use the old probe types

rather than 

- Change the semantics of the existing probe types
- Deprecate the `.Host` fields of the existing probe types
- Allow using PSA to block Pods that use probes with non-empty `.Host` fields

?

(Maybe the first case should also say "block Pods that use the old probes with non-empty `.Host` fields" rather than "block Pods that use the old probes"? I guess I was sort of abusing PSA as a way to help admins force users into the new probe semantics, but that's not really _security_ so maybe it's wrong to use PSA for that. I guess the admin could write their own admission controller if they wanted to fully deprecate the old probes.)

Reply:
> Allow using PSA to block Pods that use the old probe types

PSA or simply `ValidatingAdmissionPolicy` - it should solve pretty much all such "don't use this field" use-cases.

If we are going to do probes from in-pod network, then I think yes.  Depending on timeline, we might still want to deprecate `.Host` and recommend `VAP` for admins who want to disallow `.Host` right away.

Reply:
I have started https://github.com/kubernetes/kubernetes/pull/125271 to use PSA for "don't use this field" be careful part.